teps/0002-custom-tasks.md

```diff
@@ -0,0 +1,392 @@
+
+The Tekton Dashboard could likewise provide support for a specific well-known approval type that presents a UI to authorized users to grant or deny approval, and/or grow a plugin model that allows Custom Task authors to provide UI plugins.
+
+Both of these are out-of-scope for this proposal, but should be considered in the future.
+
+
+### Results API
+
+As a `PipelineRun` progresses, it can report status updates to a [Results API](https://docs.google.com/document/d/1-XBYQ4kBlCHIHSVoYAAf_iC01_by_KoK2aRVO0t8ZQ0/edit) ingestor endpoint. Statuses and results from a Custom Task are treated no differently. This means that the Results API should be able to support arbitrary status shapes, both when ingesting, and when serving queries.
+
+For example, a user might want to query for `PipelineRun`s that were cancelled before long wait periods completed, or where approval was denied, and even potentially by whom it was denied.
+
+### Changes to Triggers
+
+This design doesn't require any changes to the Triggers project. `PipelineRun`s or `TaskRun`s created by triggering event listeners using `TriggerTemplate`s could specify params that might change the behavior of custom tasks, just as they can change the behavior of built-in Tekton `Task`s. We might consider expanding Triggers to be able to instantiate `Run`s from `TriggerTemplate`s, as we can today with `TaskRun`s and `PipelineRun`s.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (optional)
+
+#### Task Author
+
+A user wants their pipeline to take some custom action at a certain point in their Pipeline, which does not map cleanly to a Pod execution -- for example, block further tasks on some manual approval signal.
+
+The user could simulate this behavior today by configuring a Task with a step that runs a container indefinitely until some signal is received. This could work, but incurs a performance penalty by requiring at least one container to run indefinitely in the cluster to await the signal. This also harms reliability, since the container that's waiting indefinitely might be interrupted by an underlying node failure.
```

Comment:
nit : Another way to do that would be to split the pipeline in two, using an "approval" service in the middle. If the user uses knative/serving, there might not even be any need for a container to run indefinitely. But this is then at the cost of the usability (and verbosity) of the pipeline definiton.

Reply:
Yeah, this proposal has come up before, but I don't know of anybody doing it successfully/happily today, and honeslty it seems like kind of a mess. You'd want  some way to tie together the first and second pipeline so they can be displayed together in a UI, which gets messy.

If it's a single Pipeline definition, with a "wait for approval" stage clearly defined in the middle, it becomes a _lot_ easier to display and reason about IMO.